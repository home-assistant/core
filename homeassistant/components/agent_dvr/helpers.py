"""Helpers for Agent DVR component."""


def generate_url(host, port) -> str:
    """Create a URL from the host and port."""
    server_origin = host
    if "://" not in host:
        server_origin = f"http://{host}"

    if server_origin[-1] == "/":
        server_origin = server_origin[:-1]

    return f"{server_origin}:{port}/"
