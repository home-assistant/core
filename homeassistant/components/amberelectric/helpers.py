"""Formatting helpers used to convert things."""

from amberelectric.models.price_descriptor import PriceDescriptor

DESCRIPTOR_MAP: dict[str, str] = {
    PriceDescriptor.SPIKE: "spike",
    PriceDescriptor.HIGH: "high",
    PriceDescriptor.NEUTRAL: "neutral",
    PriceDescriptor.LOW: "low",
    PriceDescriptor.VERYLOW: "very_low",
    PriceDescriptor.EXTREMELYLOW: "extremely_low",
    PriceDescriptor.NEGATIVE: "negative",
}


def normalize_descriptor(descriptor: PriceDescriptor | None) -> str | None:
    """Return the snake case versions of descriptor names. Returns None if the name is not recognized."""
    if descriptor in DESCRIPTOR_MAP:
        return DESCRIPTOR_MAP[descriptor]
    return None


def format_cents_to_dollars(cents: float) -> float:
    """Return a formatted conversion from cents to dollars."""
    return round(cents / 100, 2)
