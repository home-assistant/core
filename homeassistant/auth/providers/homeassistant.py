"""Home Assistant auth provider."""

from __future__ import annotations

import asyncio
import base64
from collections.abc import Mapping
import logging
from typing import Any, cast

import bcrypt
import voluptuous as vol

from homeassistant.const import CONF_ID
from homeassistant.core import HomeAssistant, callback
from homeassistant.exceptions import HomeAssistantError
from homeassistant.helpers import issue_registry as ir
from homeassistant.helpers.storage import Store

from ..models import AuthFlowResult, Credentials, UserMeta
from . import AUTH_PROVIDER_SCHEMA, AUTH_PROVIDERS, AuthProvider, LoginFlow

STORAGE_VERSION = 1
STORAGE_KEY = "auth_provider.homeassistant"


def _disallow_id(conf: dict[str, Any]) -> dict[str, Any]:
    """Disallow ID in config."""
    if CONF_ID in conf:
        raise vol.Invalid("ID is not allowed for the homeassistant auth provider.")

    return conf


CONFIG_SCHEMA = vol.All(AUTH_PROVIDER_SCHEMA, _disallow_id)


@callback
def async_get_provider(hass: HomeAssistant) -> HassAuthProvider:
    """Get the provider."""
    for prv in hass.auth.auth_providers:
        if prv.type == "homeassistant":
            return cast(HassAuthProvider, prv)

    raise RuntimeError("Provider not found")


class InvalidAuth(HomeAssistantError):
    """Raised when we encounter invalid authentication."""


class InvalidUser(HomeAssistantError):
    """Raised when invalid user is specified.

    Will not be raised when validating authentication.
    """

    def __init__(
        self,
        *args: object,
        translation_key: str | None = None,
        translation_placeholders: dict[str, str] | None = None,
    ) -> None:
        """Initialize exception."""
        super().__init__(
            *args,
            translation_domain="auth",
            translation_key=translation_key,
            translation_placeholders=translation_placeholders,
        )


class InvalidUsername(InvalidUser):
    """Raised when invalid username is specified.

    Will not be raised when validating authentication.
    """


class Data:
    """Hold the user data."""

    def __init__(self, hass: HomeAssistant) -> None:
        """Initialize the user data store."""
        self.hass = hass
        self._store = Store[dict[str, list[dict[str, str]]]](
            hass, STORAGE_VERSION, STORAGE_KEY, private=True, atomic_writes=True
        )
        self._data: dict[str, list[dict[str, str]]] | None = None
        # Legacy mode will allow usernames to start/end with whitespace
        # and will compare usernames case-insensitive.
        # Deprecated in June 2019 and will be removed in 2026.7
        self.is_legacy = False

    @callback
    def normalize_username(
        self, username: str, *, force_normalize: bool = False
    ) -> str:
        """Normalize a username based on the mode."""
        if self.is_legacy and not force_normalize:
            return username

        return username.strip().casefold()

    async def async_load(self) -> None:
        """Load stored data."""
        if (data := await self._store.async_load()) is None:
            data = cast(dict[str, list[dict[str, str]]], {"users": []})

        self._async_check_for_not_normalized_usernames(data)
        self._data = data

    @callback
    def _async_check_for_not_normalized_usernames(
        self, data: dict[str, list[dict[str, str]]]
    ) -> None:
        not_normalized_usernames: set[str] = set()

        for user in data["users"]:
            username = user["username"]

            if self.normalize_username(username, force_normalize=True) != username:
                logging.getLogger(__name__).warning(
                    (
                        "Home Assistant auth provider is running in legacy mode "
                        "because we detected usernames that are normalized (lowercase and without spaces)."
                        " Please change the username: '%s'."
                    ),
                    username,
                )
                not_normalized_usernames.add(username)

        if not_normalized_usernames:
            self.is_legacy = True
            ir.async_create_issue(
                self.hass,
                "auth",
                "homeassistant_provider_not_normalized_usernames",
                breaks_in_ha_version="2026.7.0",
                is_fixable=False,
                severity=ir.IssueSeverity.WARNING,
                translation_key="homeassistant_provider_not_normalized_usernames",
                translation_placeholders={
                    "usernames": f'- "{'"\n- "'.join(sorted(not_normalized_usernames))}"'
                },
                learn_more_url="homeassistant://config/users",
            )
        else:
            self.is_legacy = False
            ir.async_delete_issue(
                self.hass, "auth", "homeassistant_provider_not_normalized_usernames"
            )

    @property
    def users(self) -> list[dict[str, str]]:
        """Return users."""
        assert self._data is not None
        return self._data["users"]

    def validate_login(self, username: str, password: str) -> None:
        """Validate a username and password.

        Raises InvalidAuth if auth invalid.
        """
        username = self.normalize_username(username)
        dummy = b"$2b$12$CiuFGszHx9eNHxPuQcwBWez4CwDTOcLTX5CbOpV6gef2nYuXkY7BO"
        found = None

        # Compare all users to avoid timing attacks.
        for user in self.users:
            if self.normalize_username(user["username"]) == username:
                found = user

        if found is None:
            # check a hash to make timing the same as if user was found
            bcrypt.checkpw(b"foo", dummy)
            raise InvalidAuth

        user_hash = base64.b64decode(found["password"])

        # bcrypt.checkpw is timing-safe
        if not bcrypt.checkpw(password.encode(), user_hash):
            raise InvalidAuth

    def hash_password(self, password: str, for_storage: bool = False) -> bytes:
        """Encode a password."""
        hashed: bytes = bcrypt.hashpw(password.encode(), bcrypt.gensalt(rounds=12))

        if for_storage:
            hashed = base64.b64encode(hashed)
        return hashed

    def add_auth(self, username: str, password: str) -> None:
        """Add a new authenticated user/pass.

        Raises InvalidUsername if the new username is invalid.
        """
        self._validate_new_username(username)

        self.users.append(
            {
                "username": username,
                "password": self.hash_password(password, True).decode(),
            }
        )

    @callback
    def async_remove_auth(self, username: str) -> None:
        """Remove authentication."""
        username = self.normalize_username(username)

        index = None
        for i, user in enumerate(self.users):
            if self.normalize_username(user["username"]) == username:
                index = i
                break

        if index is None:
            raise InvalidUser(translation_key="user_not_found")

        self.users.pop(index)

    def change_password(self, username: str, new_password: str) -> None:
        """Update the password.

        Raises InvalidUser if user cannot be found.
        """
        username = self.normalize_username(username)

        for user in self.users:
            if self.normalize_username(user["username"]) == username:
                user["password"] = self.hash_password(new_password, True).decode()
                break
        else:
            raise InvalidUser(translation_key="user_not_found")

    @callback
    def _validate_new_username(self, new_username: str) -> None:
        """Validate that username is normalized and unique.

        Raises InvalidUsername if the new username is invalid.
        """
        normalized_username = self.normalize_username(
            new_username, force_normalize=True
        )
        if normalized_username != new_username:
            raise InvalidUsername(
                translation_key="username_not_normalized",
                translation_placeholders={"new_username": new_username},
            )

        if any(
            self.normalize_username(user["username"]) == normalized_username
            for user in self.users
        ):
            raise InvalidUsername(
                translation_key="username_already_exists",
                translation_placeholders={"username": new_username},
            )

    @callback
    def change_username(self, username: str, new_username: str) -> None:
        """Update the username.

        Raises InvalidUser if user cannot be found.
        Raises InvalidUsername if the new username is invalid.
        """
        username = self.normalize_username(username)
        self._validate_new_username(new_username)

        for user in self.users:
            if self.normalize_username(user["username"]) == username:
                user["username"] = new_username
                assert self._data is not None
                self._async_check_for_not_normalized_usernames(self._data)
                break
        else:
            raise InvalidUser(translation_key="user_not_found")

    async def async_save(self) -> None:
        """Save data."""
        if self._data is not None:
            await self._store.async_save(self._data)


@AUTH_PROVIDERS.register("homeassistant")
class HassAuthProvider(AuthProvider):
    """Auth provider based on a local storage of users in Home Assistant config dir."""

    DEFAULT_TITLE = "Home Assistant Local"

    def __init__(self, *args: Any, **kwargs: Any) -> None:
        """Initialize an Home Assistant auth provider."""
        super().__init__(*args, **kwargs)
        self.data: Data | None = None
        self._init_lock = asyncio.Lock()

    async def async_initialize(self) -> None:
        """Initialize the auth provider."""
        async with self._init_lock:
            if self.data is not None:
                return

            data = Data(self.hass)
            await data.async_load()
            self.data = data

    async def async_login_flow(self, context: dict[str, Any] | None) -> LoginFlow:
        """Return a flow to login."""
        return HassLoginFlow(self)

    async def async_validate_login(self, username: str, password: str) -> None:
        """Validate a username and password."""
        if self.data is None:
            await self.async_initialize()
            assert self.data is not None

        await self.hass.async_add_executor_job(
            self.data.validate_login, username, password
        )

    async def async_add_auth(self, username: str, password: str) -> None:
        """Call add_auth on data."""
        if self.data is None:
            await self.async_initialize()
            assert self.data is not None

        await self.hass.async_add_executor_job(self.data.add_auth, username, password)
        await self.data.async_save()

    async def async_remove_auth(self, username: str) -> None:
        """Call remove_auth on data."""
        if self.data is None:
            await self.async_initialize()
            assert self.data is not None

        self.data.async_remove_auth(username)
        await self.data.async_save()

    async def async_change_password(self, username: str, new_password: str) -> None:
        """Call change_password on data."""
        if self.data is None:
            await self.async_initialize()
            assert self.data is not None

        await self.hass.async_add_executor_job(
            self.data.change_password, username, new_password
        )
        await self.data.async_save()

    async def async_change_username(
        self, credential: Credentials, new_username: str
    ) -> None:
        """Validate new username and change it including updating credentials object."""
        if self.data is None:
            await self.async_initialize()
            assert self.data is not None

        self.data.change_username(credential.data["username"], new_username)
        self.hass.auth.async_update_user_credentials_data(
            credential, {**credential.data, "username": new_username}
        )
        await self.data.async_save()

    async def async_get_or_create_credentials(
        self, flow_result: Mapping[str, str]
    ) -> Credentials:
        """Get credentials based on the flow result."""
        if self.data is None:
            await self.async_initialize()
            assert self.data is not None

        norm_username = self.data.normalize_username
        username = norm_username(flow_result["username"])

        for credential in await self.async_credentials():
            if norm_username(credential.data["username"]) == username:
                return credential

        # Create new credentials.
        return self.async_create_credentials({"username": username})

    async def async_user_meta_for_credentials(
        self, credentials: Credentials
    ) -> UserMeta:
        """Get extra info for this credential."""
        return UserMeta(name=credentials.data["username"], is_active=True)

    async def async_will_remove_credentials(self, credentials: Credentials) -> None:
        """When credentials get removed, also remove the auth."""
        if self.data is None:
            await self.async_initialize()
            assert self.data is not None

        try:
            self.data.async_remove_auth(credentials.data["username"])
            await self.data.async_save()
        except InvalidUser:
            # Can happen if somehow we didn't clean up a credential
            pass


class HassLoginFlow(LoginFlow):
    """Handler for the login flow."""

    async def async_step_init(
        self, user_input: dict[str, str] | None = None
    ) -> AuthFlowResult:
        """Handle the step of the form."""
        errors = {}

        if user_input is not None:
            try:
                await cast(HassAuthProvider, self._auth_provider).async_validate_login(
                    user_input["username"], user_input["password"]
                )
            except InvalidAuth:
                errors["base"] = "invalid_auth"

            if not errors:
                user_input.pop("password")
                return await self.async_finish(user_input)

        return self.async_show_form(
            step_id="init",
            data_schema=vol.Schema(
                {
                    vol.Required("username"): str,
                    vol.Required("password"): str,
                }
            ),
            errors=errors,
        )
