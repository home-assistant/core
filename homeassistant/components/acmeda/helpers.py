"""Helper functions for Acmeda Pulse."""

from __future__ import annotations

from typing import TYPE_CHECKING

from aiopulse import Roller

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers import device_registry as dr
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .const import DOMAIN, LOGGER

if TYPE_CHECKING:
    from . import AcmedaConfigEntry


@callback
def async_add_acmeda_entities(
    hass: HomeAssistant,
    entity_class: type,
    config_entry: AcmedaConfigEntry,
    current: set[int],
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Add any new entities."""
    hub = config_entry.runtime_data
    LOGGER.debug("Looking for new %s on: %s", entity_class.__name__, hub.host)

    api = hub.api.rollers

    new_items = []
    for unique_id, roller in api.items():
        if unique_id not in current:
            LOGGER.debug("New %s %s", entity_class.__name__, unique_id)
            new_item = entity_class(roller)
            current.add(unique_id)
            new_items.append(new_item)

    async_add_entities(new_items)


async def update_devices(
    hass: HomeAssistant, config_entry: ConfigEntry, api: dict[int, Roller]
) -> None:
    """Tell hass that device info has been updated."""
    dev_registry = dr.async_get(hass)

    for api_item in api.values():
        # Update Device name
        device = dev_registry.async_get_device(identifiers={(DOMAIN, api_item.id)})
        if device is not None:
            dev_registry.async_update_device(
                device.id,
                name=api_item.name,
            )
