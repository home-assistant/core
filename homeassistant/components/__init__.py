"""Contains components that can be plugged into Home Assistant.

Component design guidelines:
- Each component defines a constant DOMAIN that is equal to its filename.
- Each component that tracks states should create state entity names in the
  format "<DOMAIN>.<OBJECT_ID>".
- Each component should publish services only under its own domain.
"""

from __future__ import annotations

import logging

from homeassistant.core import HomeAssistant, split_entity_id
from homeassistant.helpers.frame import report
from homeassistant.helpers.group import expand_entity_ids

_LOGGER = logging.getLogger(__name__)


def is_on(hass: HomeAssistant, entity_id: str | None = None) -> bool:
    """Load up the module to call the is_on method.

    If there is no entity id given we will check all.
    """
    report(
        (
            "uses homeassistant.components.is_on."
            " This is deprecated and will stop working in Home Assistant 2024.9, it"
            " should be updated to use the function of the platform directly."
        ),
        error_if_core=True,
    )

    if entity_id:
        entity_ids = expand_entity_ids(hass, [entity_id])
    else:
        entity_ids = hass.states.entity_ids()

    for ent_id in entity_ids:
        domain = split_entity_id(ent_id)[0]

        try:
            component = getattr(hass.components, domain)

        except ImportError:
            _LOGGER.error("Failed to call %s.is_on: component not found", domain)
            continue

        if not hasattr(component, "is_on"):
            _LOGGER.warning("Integration %s has no is_on method", domain)
            continue

        if component.is_on(ent_id):
            return True

    return False
