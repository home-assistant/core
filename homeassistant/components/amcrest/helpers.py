"""Helpers for amcrest component."""

from __future__ import annotations

import logging

from homeassistant.helpers.typing import UndefinedType

from .const import DOMAIN


def service_signal(service: str, *args: str) -> str:
    """Encode signal."""
    return "_".join([DOMAIN, service, *args])


def log_update_error(
    logger: logging.Logger,
    action: str,
    name: str | UndefinedType | None,
    entity_type: str,
    error: Exception,
    level: int = logging.ERROR,
) -> None:
    """Log an update error."""
    logger.log(
        level,
        "Could not %s %s %s due to error: %s",
        action,
        name,
        entity_type,
        error.__class__.__name__,
    )
