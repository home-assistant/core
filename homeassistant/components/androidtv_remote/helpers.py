"""Helper functions for Android TV Remote integration."""

from __future__ import annotations

from androidtvremote2 import AndroidTVRemote

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.storage import STORAGE_DIR

from .const import CONF_ENABLE_IME, CONF_ENABLE_IME_DEFAULT_VALUE


def create_api(hass: HomeAssistant, host: str, enable_ime: bool) -> AndroidTVRemote:
    """Create an AndroidTVRemote instance."""
    return AndroidTVRemote(
        client_name="Home Assistant",
        certfile=hass.config.path(STORAGE_DIR, "androidtv_remote_cert.pem"),
        keyfile=hass.config.path(STORAGE_DIR, "androidtv_remote_key.pem"),
        host=host,
        loop=hass.loop,
        enable_ime=enable_ime,
    )


def get_enable_ime(entry: ConfigEntry) -> bool:
    """Get value of enable_ime option or its default value."""
    return entry.options.get(CONF_ENABLE_IME, CONF_ENABLE_IME_DEFAULT_VALUE)
