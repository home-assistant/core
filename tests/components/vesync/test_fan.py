"""Tests for the fan module."""

import pytest
import requests_mock
from syrupy import SnapshotAssertion

from homeassistant.components.fan import DOMAIN as FAN_DOMAIN
from homeassistant.core import HomeAssistant
from homeassistant.helpers import device_registry as dr, entity_registry as er

from .common import ALL_DEVICE_NAMES, mock_devices_response

from tests.common import MockConfigEntry


@pytest.mark.parametrize("device_name", ALL_DEVICE_NAMES)
async def test_fan_state(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    config_entry: MockConfigEntry,
    device_registry: dr.DeviceRegistry,
    entity_registry: er.EntityRegistry,
    requests_mock: requests_mock.Mocker,
    device_name: str,
) -> None:
    """Test the resulting setup state is as expected for the platform."""

    # Configure the API devices call for device_name
    mock_devices_response(requests_mock, device_name)

    # setup platform - only including the named device
    await hass.config_entries.async_setup(config_entry.entry_id)
    await hass.async_block_till_done()

    # Check device registry
    devices = dr.async_entries_for_config_entry(device_registry, config_entry.entry_id)
    assert devices == snapshot(name="devices")

    # Check entity registry
    entities = [
        entity
        for entity in er.async_entries_for_config_entry(
            entity_registry, config_entry.entry_id
        )
        if entity.domain == FAN_DOMAIN
    ]
    assert entities == snapshot(name="entities")

    # Check states
    for entity in entities:
        assert hass.states.get(entity.entity_id) == snapshot(name=entity.entity_id)
