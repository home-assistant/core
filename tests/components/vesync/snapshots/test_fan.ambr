# serializer version: 1
# name: test_fan_state[Air Purifier 131s][devices]
  list([
    DeviceRegistryEntrySnapshot({
      'area_id': None,
      'config_entries': <ANY>,
      'configuration_url': None,
      'connections': set({
      }),
      'disabled_by': None,
      'entry_type': None,
      'hw_version': None,
      'id': <ANY>,
      'identifiers': set({
        tuple(
          'vesync',
          'air-purifier',
        ),
      }),
      'is_new': False,
      'labels': set({
      }),
      'manufacturer': 'VeSync',
      'model': 'LV-PUR131S',
      'name': 'Air Purifier 131s',
      'name_by_user': None,
      'primary_config_entry': <ANY>,
      'serial_number': None,
      'suggested_area': None,
      'sw_version': None,
      'via_device_id': None,
    }),
  ])
# ---
# name: test_fan_state[Air Purifier 131s][entities]
  list([
    EntityRegistryEntrySnapshot({
      'aliases': set({
      }),
      'area_id': None,
      'capabilities': dict({
        'preset_modes': list([
          'auto',
          'sleep',
        ]),
      }),
      'config_entry_id': <ANY>,
      'device_class': None,
      'device_id': <ANY>,
      'disabled_by': None,
      'domain': 'fan',
      'entity_category': None,
      'entity_id': 'fan.air_purifier_131s',
      'has_entity_name': True,
      'hidden_by': None,
      'icon': None,
      'id': <ANY>,
      'labels': set({
      }),
      'name': None,
      'options': dict({
      }),
      'original_device_class': None,
      'original_icon': None,
      'original_name': None,
      'platform': 'vesync',
      'previous_unique_id': None,
      'supported_features': <FanEntityFeature: 9>,
      'translation_key': None,
      'unique_id': 'air-purifier',
      'unit_of_measurement': None,
    }),
  ])
# ---
# name: test_fan_state[Air Purifier 131s][fan.air_purifier_131s]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Air Purifier 131s',
      'preset_modes': list([
        'auto',
        'sleep',
      ]),
      'supported_features': <FanEntityFeature: 9>,
    }),
    'context': <ANY>,
    'entity_id': 'fan.air_purifier_131s',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unavailable',
  })
# ---
# name: test_fan_state[Air Purifier 200s][devices]
  list([
    DeviceRegistryEntrySnapshot({
      'area_id': None,
      'config_entries': <ANY>,
      'configuration_url': None,
      'connections': set({
      }),
      'disabled_by': None,
      'entry_type': None,
      'hw_version': None,
      'id': <ANY>,
      'identifiers': set({
        tuple(
          'vesync',
          'asd_sdfKIHG7IJHGwJGJ7GJ_ag5h3G55',
        ),
      }),
      'is_new': False,
      'labels': set({
      }),
      'manufacturer': 'VeSync',
      'model': 'Core200S',
      'name': 'Air Purifier 200s',
      'name_by_user': None,
      'primary_config_entry': <ANY>,
      'serial_number': None,
      'suggested_area': None,
      'sw_version': None,
      'via_device_id': None,
    }),
  ])
# ---
# name: test_fan_state[Air Purifier 200s][entities]
  list([
    EntityRegistryEntrySnapshot({
      'aliases': set({
      }),
      'area_id': None,
      'capabilities': dict({
        'preset_modes': list([
          'sleep',
        ]),
      }),
      'config_entry_id': <ANY>,
      'device_class': None,
      'device_id': <ANY>,
      'disabled_by': None,
      'domain': 'fan',
      'entity_category': None,
      'entity_id': 'fan.air_purifier_200s',
      'has_entity_name': True,
      'hidden_by': None,
      'icon': None,
      'id': <ANY>,
      'labels': set({
      }),
      'name': None,
      'options': dict({
      }),
      'original_device_class': None,
      'original_icon': None,
      'original_name': None,
      'platform': 'vesync',
      'previous_unique_id': None,
      'supported_features': <FanEntityFeature: 9>,
      'translation_key': None,
      'unique_id': 'asd_sdfKIHG7IJHGwJGJ7GJ_ag5h3G55',
      'unit_of_measurement': None,
    }),
  ])
# ---
# name: test_fan_state[Air Purifier 200s][fan.air_purifier_200s]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'child_lock': False,
      'friendly_name': 'Air Purifier 200s',
      'mode': 'manual',
      'night_light': 'off',
      'percentage': 33,
      'percentage_step': 33.333333333333336,
      'preset_mode': None,
      'preset_modes': list([
        'sleep',
      ]),
      'screen_status': True,
      'supported_features': <FanEntityFeature: 9>,
    }),
    'context': <ANY>,
    'entity_id': 'fan.air_purifier_200s',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_fan_state[Air Purifier 400s][devices]
  list([
    DeviceRegistryEntrySnapshot({
      'area_id': None,
      'config_entries': <ANY>,
      'configuration_url': None,
      'connections': set({
      }),
      'disabled_by': None,
      'entry_type': None,
      'hw_version': None,
      'id': <ANY>,
      'identifiers': set({
        tuple(
          'vesync',
          '400s-purifier',
        ),
      }),
      'is_new': False,
      'labels': set({
      }),
      'manufacturer': 'VeSync',
      'model': 'LAP-C401S-WJP',
      'name': 'Air Purifier 400s',
      'name_by_user': None,
      'primary_config_entry': <ANY>,
      'serial_number': None,
      'suggested_area': None,
      'sw_version': None,
      'via_device_id': None,
    }),
  ])
# ---
# name: test_fan_state[Air Purifier 400s][entities]
  list([
    EntityRegistryEntrySnapshot({
      'aliases': set({
      }),
      'area_id': None,
      'capabilities': dict({
        'preset_modes': list([
          'auto',
          'sleep',
        ]),
      }),
      'config_entry_id': <ANY>,
      'device_class': None,
      'device_id': <ANY>,
      'disabled_by': None,
      'domain': 'fan',
      'entity_category': None,
      'entity_id': 'fan.air_purifier_400s',
      'has_entity_name': True,
      'hidden_by': None,
      'icon': None,
      'id': <ANY>,
      'labels': set({
      }),
      'name': None,
      'options': dict({
      }),
      'original_device_class': None,
      'original_icon': None,
      'original_name': None,
      'platform': 'vesync',
      'previous_unique_id': None,
      'supported_features': <FanEntityFeature: 9>,
      'translation_key': None,
      'unique_id': '400s-purifier',
      'unit_of_measurement': None,
    }),
  ])
# ---
# name: test_fan_state[Air Purifier 400s][fan.air_purifier_400s]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'child_lock': False,
      'friendly_name': 'Air Purifier 400s',
      'mode': 'manual',
      'night_light': 'off',
      'percentage': 25,
      'percentage_step': 25.0,
      'preset_mode': None,
      'preset_modes': list([
        'auto',
        'sleep',
      ]),
      'screen_status': True,
      'supported_features': <FanEntityFeature: 9>,
    }),
    'context': <ANY>,
    'entity_id': 'fan.air_purifier_400s',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_fan_state[Air Purifier 600s][devices]
  list([
    DeviceRegistryEntrySnapshot({
      'area_id': None,
      'config_entries': <ANY>,
      'configuration_url': None,
      'connections': set({
      }),
      'disabled_by': None,
      'entry_type': None,
      'hw_version': None,
      'id': <ANY>,
      'identifiers': set({
        tuple(
          'vesync',
          '600s-purifier',
        ),
      }),
      'is_new': False,
      'labels': set({
      }),
      'manufacturer': 'VeSync',
      'model': 'LAP-C601S-WUS',
      'name': 'Air Purifier 600s',
      'name_by_user': None,
      'primary_config_entry': <ANY>,
      'serial_number': None,
      'suggested_area': None,
      'sw_version': None,
      'via_device_id': None,
    }),
  ])
# ---
# name: test_fan_state[Air Purifier 600s][entities]
  list([
    EntityRegistryEntrySnapshot({
      'aliases': set({
      }),
      'area_id': None,
      'capabilities': dict({
        'preset_modes': list([
          'auto',
          'sleep',
        ]),
      }),
      'config_entry_id': <ANY>,
      'device_class': None,
      'device_id': <ANY>,
      'disabled_by': None,
      'domain': 'fan',
      'entity_category': None,
      'entity_id': 'fan.air_purifier_600s',
      'has_entity_name': True,
      'hidden_by': None,
      'icon': None,
      'id': <ANY>,
      'labels': set({
      }),
      'name': None,
      'options': dict({
      }),
      'original_device_class': None,
      'original_icon': None,
      'original_name': None,
      'platform': 'vesync',
      'previous_unique_id': None,
      'supported_features': <FanEntityFeature: 9>,
      'translation_key': None,
      'unique_id': '600s-purifier',
      'unit_of_measurement': None,
    }),
  ])
# ---
# name: test_fan_state[Air Purifier 600s][fan.air_purifier_600s]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'child_lock': False,
      'friendly_name': 'Air Purifier 600s',
      'mode': 'manual',
      'night_light': 'off',
      'percentage': 25,
      'percentage_step': 25.0,
      'preset_mode': None,
      'preset_modes': list([
        'auto',
        'sleep',
      ]),
      'screen_status': True,
      'supported_features': <FanEntityFeature: 9>,
    }),
    'context': <ANY>,
    'entity_id': 'fan.air_purifier_600s',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_fan_state[Dimmable Light][devices]
  list([
    DeviceRegistryEntrySnapshot({
      'area_id': None,
      'config_entries': <ANY>,
      'configuration_url': None,
      'connections': set({
      }),
      'disabled_by': None,
      'entry_type': None,
      'hw_version': None,
      'id': <ANY>,
      'identifiers': set({
        tuple(
          'vesync',
          'dimmable-bulb',
        ),
      }),
      'is_new': False,
      'labels': set({
      }),
      'manufacturer': 'VeSync',
      'model': 'ESL100',
      'name': 'Dimmable Light',
      'name_by_user': None,
      'primary_config_entry': <ANY>,
      'serial_number': None,
      'suggested_area': None,
      'sw_version': None,
      'via_device_id': None,
    }),
  ])
# ---
# name: test_fan_state[Dimmable Light][entities]
  list([
  ])
# ---
# name: test_fan_state[Dimmer Switch][devices]
  list([
    DeviceRegistryEntrySnapshot({
      'area_id': None,
      'config_entries': <ANY>,
      'configuration_url': None,
      'connections': set({
      }),
      'disabled_by': None,
      'entry_type': None,
      'hw_version': None,
      'id': <ANY>,
      'identifiers': set({
        tuple(
          'vesync',
          'dimmable-switch',
        ),
      }),
      'is_new': False,
      'labels': set({
      }),
      'manufacturer': 'VeSync',
      'model': 'ESWD16',
      'name': 'Dimmer Switch',
      'name_by_user': None,
      'primary_config_entry': <ANY>,
      'serial_number': None,
      'suggested_area': None,
      'sw_version': None,
      'via_device_id': None,
    }),
  ])
# ---
# name: test_fan_state[Dimmer Switch][entities]
  list([
  ])
# ---
# name: test_fan_state[Humidifier 200s][devices]
  list([
  ])
# ---
# name: test_fan_state[Humidifier 200s][entities]
  list([
  ])
# ---
# name: test_fan_state[Humidifier 600S][devices]
  list([
  ])
# ---
# name: test_fan_state[Humidifier 600S][entities]
  list([
  ])
# ---
# name: test_fan_state[Outlet][devices]
  list([
    DeviceRegistryEntrySnapshot({
      'area_id': None,
      'config_entries': <ANY>,
      'configuration_url': None,
      'connections': set({
      }),
      'disabled_by': None,
      'entry_type': None,
      'hw_version': None,
      'id': <ANY>,
      'identifiers': set({
        tuple(
          'vesync',
          'outlet',
        ),
      }),
      'is_new': False,
      'labels': set({
      }),
      'manufacturer': 'VeSync',
      'model': 'wifi-switch-1.3',
      'name': 'Outlet',
      'name_by_user': None,
      'primary_config_entry': <ANY>,
      'serial_number': None,
      'suggested_area': None,
      'sw_version': None,
      'via_device_id': None,
    }),
  ])
# ---
# name: test_fan_state[Outlet][entities]
  list([
  ])
# ---
# name: test_fan_state[Temperature Light][devices]
  list([
    DeviceRegistryEntrySnapshot({
      'area_id': None,
      'config_entries': <ANY>,
      'configuration_url': None,
      'connections': set({
      }),
      'disabled_by': None,
      'entry_type': None,
      'hw_version': None,
      'id': <ANY>,
      'identifiers': set({
        tuple(
          'vesync',
          'tunable-bulb',
        ),
      }),
      'is_new': False,
      'labels': set({
      }),
      'manufacturer': 'VeSync',
      'model': 'ESL100CW',
      'name': 'Temperature Light',
      'name_by_user': None,
      'primary_config_entry': <ANY>,
      'serial_number': None,
      'suggested_area': None,
      'sw_version': None,
      'via_device_id': None,
    }),
  ])
# ---
# name: test_fan_state[Temperature Light][entities]
  list([
  ])
# ---
# name: test_fan_state[Wall Switch][devices]
  list([
    DeviceRegistryEntrySnapshot({
      'area_id': None,
      'config_entries': <ANY>,
      'configuration_url': None,
      'connections': set({
      }),
      'disabled_by': None,
      'entry_type': None,
      'hw_version': None,
      'id': <ANY>,
      'identifiers': set({
        tuple(
          'vesync',
          'switch',
        ),
      }),
      'is_new': False,
      'labels': set({
      }),
      'manufacturer': 'VeSync',
      'model': 'ESWL01',
      'name': 'Wall Switch',
      'name_by_user': None,
      'primary_config_entry': <ANY>,
      'serial_number': None,
      'suggested_area': None,
      'sw_version': None,
      'via_device_id': None,
    }),
  ])
# ---
# name: test_fan_state[Wall Switch][entities]
  list([
  ])
# ---
